plugins/fetchinvoice.c

```diff
@@ -851,6 +867,54 @@ static struct command_result *json_fetchinvoice(struct command *cmd,
 	return send_outreq(cmd->plugin, req);
 }
 
+/* FIXME: Using a hook here is not ideal: technically it doesn't mean
+ * it's actually hit the db!  But using waitinvoice is also suboptimal
+ * because we don't have libplugin infra to cancel a pending req (and I
+ * want to rewrite our wait* API anyway) */
+static struct command_result *invoice_payment(struct command *cmd,
+					      const char *buf,
+					      const jsmntok_t *params)
+{
+	struct sent *i;
+	const jsmntok_t *ptok, *preimagetok, *msattok;
+	struct preimage preimage;
+	struct amount_msat msat;
+
+	ptok = json_get_member(buf, params, "payment");
+	preimagetok = json_get_member(buf, ptok, "preimage");
+	msattok = json_get_member(buf, ptok, "msat");
+	if (!preimagetok || !msattok)
+		plugin_err(cmd->plugin,
+			   "Invalid invoice_payment %.*s",
+			   json_tok_full_len(params),
+			   json_tok_full(buf, params));
+
+	hex_decode(buf + preimagetok->start,
+		   preimagetok->end - preimagetok->start,
+		   &preimage, sizeof(preimage));
+	json_to_msat(buf, msattok, &msat);
+
+	list_for_each(&sent_list, i, list) {
+		struct json_stream *out;
+
+		if (!i->inv)
+			continue;
+		if (!preimage_eq(&preimage, &i->inv_preimage))
+			continue;
+
+		/* It was paid!  Success. */
+		/* FIXME: Return as per waitinvoice */
```

Comment:
We could just call `waitinvoice` since that's supposed to return immediately on paid invoices, and it'd allow us to avoid copy-pasting details here.

Reply:
Ah, good idea!